exporter/diagnostic_data_collector.go

```diff
@@ -66,39 +66,46 @@ func (d *diagnosticDataCollector) collect(ch chan<- prometheus.Metric) {
 	logger := d.base.logger
 	client := d.base.client
 
+	nodeType, err := getNodeType(d.ctx, client)
+	if err != nil {
+		logger.WithFields(logrus.Fields{
+			"component": "diagnosticDataCollector",
+		}).Errorf("Cannot get node type: %s", err)
+	}
+
+	var metrics []prometheus.Metric
 	cmd := bson.D{{Key: "getDiagnosticData", Value: "1"}}
 	res := client.Database("admin").RunCommand(d.ctx, cmd)
 	if res.Err() != nil {
-		logger.Errorf("failed to run command: getDiagnosticData: %s", res.Err())
-		logger.Warn("cannot run getDiagnosticData, some metrics might be unavailable.")
-	}
-
-	if err := res.Decode(&m); err != nil {
-		logger.Errorf("cannot run getDiagnosticData: %s", err)
-	}
-
-	if m == nil || m["data"] == nil {
-		logger.Error("cannot run getDiagnosticData: response is empty")
-	}
+		logger.Warnf("failed to run command: getDiagnosticData, some metrics might be unavailable %s", res.Err())
+	} else {
+		if err := res.Decode(&m); err != nil {
+			logger.Errorf("cannot run getDiagnosticData: %s", err)
+		}
 
-	m, ok := m["data"].(bson.M)
-	if !ok {
-		err := errors.Wrapf(errUnexpectedDataType, "%T for data field", m["data"])
+		if m == nil || m["data"] == nil {
+			logger.Error("cannot run getDiagnosticData: response is empty")
+		}
 
-		logger.Errorf("cannot decode getDiagnosticData: %s", err)
-	}
+		var ok bool
+		m, ok = m["data"].(bson.M)
+		if !ok {
+			err = errors.Wrapf(errUnexpectedDataType, "%T for data field", m["data"])
+			logger.Errorf("cannot decode getDiagnosticData: %s", err)
+		}
 
-	logger.Debug("getDiagnosticData result")
-	debugResult(logger, m)
+		logger.Debug("getDiagnosticData result")
+		debugResult(logger, m)
 
-	metrics := makeMetrics("", m, d.topologyInfo.baseLabels(), d.compatibleMode)
-	metrics = append(metrics, locksMetrics(logger, m)...)
+		metrics = makeMetrics("", m, d.topologyInfo.baseLabels(), d.compatibleMode)
+		metrics = append(metrics, locksMetrics(logger, m)...)
 
-	securityMetric, err := d.getSecurityMetricFromLineOptions(client)
-	if err != nil {
-		logger.Errorf("cannot decode getCmdLineOtpions: %s", err)
-	} else if securityMetric != nil {
-		metrics = append(metrics, securityMetric)
+		securityMetric, err := d.getSecurityMetricFromLineOptions(client)
+		if err != nil {
+			logger.Errorf("failed to run command: getCmdLineOptions: %s", err)
+		} else if securityMetric != nil {
+			metrics = append(metrics, securityMetric)
+		}
 	}
 
 	if d.compatibleMode {
```

Comment:
at this point `m` might be empty, so it makes not much sense to call lines https://github.com/percona/mongodb_exporter/pull/820/files#diff-f08bb8e2230b960e7f7941dbdf020e3aa607391de1dc0f8683c315265d77b2f6R112-R116
I would include them to else part of the statement above

Reply:
BTW can we get diagnostic data from mongos?

Reply:
> BTW can we get diagnostic data from mongos?

No actually, the command won't fail, but `m["data"]` will be empty.